lib/matrix.rb

```diff
@@ -303,12 +303,139 @@ def [](i, j)
   alias element []
   alias component []
 
+  #
+  # :call-seq:
+  #   matrix[range, range] = new_matrix
+  #   matrix[range, range] = new_element
+  #   matrix[range, integer] = vector
+  #   matrix[range, integer] = column_matrix
+  #   matrix[range, integer] = new_element
+  #   matrix[integer, range] = vector
+  #   matrix[integer, range] = row_matrix
+  #   matrix[integer, range] = new_element
+  #   matrix[integer, integer] = new_element
+  #
+  # Set element or elements of matrix.
   def []=(i, j, v)
-    @rows[i][j] = v
+    raise FrozenError, "can't modify frozen Matrix" if frozen?
+    if both_ranges?(i, j)
+      i = convert_range(i, row_count) if i.begin > i.end
+      j = convert_range(j, column_count) if j.begin > j.end
+      raise IndexError, "expected ranges are outside of matrix" unless range_within_matrix_range?(i, j)
+      set_row_and_col_range(i, j, v)
+    elsif one_is_range?(i, j)
+      set_range(i, j, v)
+    else
+      i = CoercionHelper.coerce_to_int(i)
+      j = CoercionHelper.coerce_to_int(j)
+      raise IndexError, "indices are outside of matrix" unless indices_within_matrix?(i, j)
+      set_value(i, j, v)
+    end
   end
   alias set_element []=
   alias set_component []=
-  private :[]=, :set_element, :set_component
+
+  def range_within_count?(range, count)
+    (range.max < count) && (range.min >= (-count))
+  end
+
+  def range_within_matrix_range?(row_range, col_range)
+    range_within_count?(row_range, row_count) && range_within_count?(col_range, column_count)
+  end
+
+  def ranges_and_dimensions_equal?(row, col, matrix)
+    row.size == matrix.row_count && col.size == matrix.column_count
+  end
+
+  private :range_within_count?, :range_within_matrix_range?, :ranges_and_dimensions_equal?
+
+  def indices_within_matrix?(row_index, col_index)
+    row_index.between?(-row_count, row_count-1) && col_index.between?(-column_count, column_count-1)
+  end
+
+  def both_ranges?(row, col)
+    row.is_a?(Range) && col.is_a?(Range)
+  end
+
+  def one_is_range?(row, col)
+    row.is_a?(Range) || col.is_a?(Range)
+  end
+
+  def convert_range(range, count)
+    range.exclude_end? ? range.begin...(count + range.end) : range.begin..(count + range.end)
+  end
+
+  private :indices_within_matrix?, :both_ranges?, :one_is_range?, :convert_range
+
+  def set_value(row, col, value)
+    @rows[row][col] = value
+  end
+
+  def set_range(row, col, value)
+    if row.is_a?(Range)
+      row = convert_range(row, row_count) if row.begin > row.end
```

Comment:
I have the feeling that it would be much better to do the checking and error raising in the same method. Error message could be made more instructive by giving the actual wrong values too.